### PR TITLE
Remove undefined behaviour for access of empty std::vectors

### PR DESCRIPTION
**Pull request by marksidoro**
Below, I have updated some places at which we currently have code of the form `&vec[0]`, which is undefined behaviour if the vector is empty, as the access operator is not defined in that case. My gcc asserts this in debug mode, that's how I came across the issue. 
My proposed solution is to use std::vector::data() plus pointer arithmetics, as data() is guaranteed to return null for an empty vector. However this takes out the assertion in debug mode completely so maybe the approach depicted in `src/core/rebalance/src/4C_rebalance_graph_based.cpp` is more suitable.

### CODE REVIEW

File: src/core/comm/src/4C_comm_pack_buffer.hpp [hunk 1]
```diff
@@ -137,16 +137,17 @@ namespace Core::Communication
...
           "Type mismatch during unpacking. Tried to extract type {}", typeid(T).name());
 #endif
 
-
+      FOUR_C_ASSERT(position_ + stuff_size <= data_.size(),
+          "UnpackBuffer: Trying to extract more data than available in the buffer.");
       // We know that stuff is trivially copyable, so we can safely use memcpy. We cast to void* to
       // circumvent a GCC warning
-      std::memcpy(static_cast<void*>(stuff), &data_[position_], stuff_size);
+      memcpy(static_cast<void*>(stuff), data_.data() + position_, stuff_size);
```
Comment (sebproell) [inline src/core/comm/src/4C_comm_pack_buffer.hpp]:
```suggestion
      std::memcpy(static_cast<void*>(stuff), data_.data() + position_, stuff_size);
```

File: src/core/comm/src/4C_comm_pack_buffer.hpp [hunk 2]
```diff
@@ -162,12 +163,13 @@ namespace Core::Communication
...
-      std::memcpy(&hash, &data_[position_], sizeof(hash));
+      std::memcpy(&hash, data_.data() + position_, sizeof(hash));
       position += sizeof(hash);
       FOUR_C_ASSERT(hash == typeid(T).hash_code(), "Type mismatch during unpacking.");
 #endif
-
-      memcpy(&stuff, &data_[position], sizeof(T));
+      FOUR_C_ASSERT(position < data_.size(),
+          "UnpackBuffer: Trying to extract more data than available in the buffer.");
+      memcpy(&stuff, data_.data() + position, sizeof(T));
```
Comment (sebproell) [inline src/core/comm/src/4C_comm_pack_buffer.hpp]:
```suggestion
      std::memcpy(&stuff, data_.data() + position, sizeof(T));
```

**Comment by sebproell** [comment]
> However this takes out the assertion in debug mode completely so maybe the approach depicted in src/core/rebalance/src/4C_rebalance_graph_based.cpp is more suitable.

I don't understand this. What does 

>  takes out the assertion in debug mode completely 

mean?

**Review by sebproell — COMMENTED** [review]
Can you add a test for the empty vector case  in `4C_comm_pack_unpack_test.np3.cpp`?

**Comment by marksidoro** [comment]
Currently, the vector performs an index check during the `operator[](size_t i)` call i.e. if i < vector.size(). Using the data() + i approach this Assertion is not done. 
But during a call like  `vector.empty() ? null_ptr :   &vector[i]` the assertion would remain.

File: src/core/rebalance/src/4C_rebalance_graph_based.cpp [hunk 3]
```diff
@@ -301,7 +301,8 @@ std::shared_ptr<const Core::LinAlg::Graph> Core::Rebalance::build_graph(
     }
     Core::Communication::broadcast(&size, 1, proc, dis.get_comm());
     if (proc != myrank) recvnodes.resize(size);
-    Core::Communication::broadcast(&recvnodes[0], size, proc, dis.get_comm());
+    Core::Communication::broadcast(
+        recvnodes.empty() ? nullptr : &recvnodes[0], size, proc, dis.get_comm());
```
Comment (amgebauer) [inline src/core/rebalance/src/4C_rebalance_graph_based.cpp]: Is it allowed to broadcast the nullptr?


```suggestion
    if (!recvnodes.empty()) Core::Communication::broadcast(&recvnodes[0], size, proc, dis.get_comm());
```

**Review by amgebauer — COMMENTED** [review]
> My proposed solution is to use std::vector::data() plus pointer arithmetics, as data() is guaranteed to return null for an empty vector.

Isn't that still undefined behavior? Then you are adding something to the nullptr

**Comment by marksidoro** [comment]
> Isn't that still undefined behavior? Then you are adding something to the nullptr

Not if we add zero as `nullptr +/- 0` is well defined. 
For i =nonzero case it is indeed undefined but so is vec[i] in the first place, so we don't "lose" anything

**Comment by marksidoro** [comment]
After talking to @sebproell we decided to check the `stuff_size` int to make sure we only call `std::memcpy` if we should have valid vectors. 
I also did a global search and updated all places where we used `&vec[0]` and vec could feasibly be empty to use vec.data() instead

**Review by amgebauer — APPROVED** [review]
